File: Array/875.c
#include <stdio.h>

int minEatingSpeed(int* piles, int pilesSize, int h){

    //最慢1根/1小時
    //最快就是max(piles[i])
    int left = 1;
    int right = piles[0]; //max of piles[i]
    for(int i = 1; i < pilesSize; i++){
        if(piles[i]> right){
            right = piles[i];
        }
    }
    int k;
    int ans;
    while(left <= right){
        k = (left+right)/2;
        long long hours = 0;
        for(int i = 0; i < pilesSize; i++){
            //hours += piles[i] / k+1;
            //不行的原因是因為整除的時候不需要加1
            //除之前先多給它「k-1 的緩衝」
            //只要有剩，就一定會被推到下一個整數
            hours+= (piles[i] + k - 1) / k;
        }
        if(hours > h){ //太慢
            left = k+1;
        }else{//太快
            right = k-1;
            ans = k;
        }
    }

    return ans;

}

static void test(int* piles, int n, int h, int expect) {
    int ans = minEatingSpeed(piles, n, h);
    printf("ans=%d expect=%d %s\n", ans, expect, ans==expect?"OK":"WRONG");
}

int main() {
    int a1[] = {3,6,7,11};
    test(a1, 4, 8, 4);

    int a2[] = {30,11,23,4,20};
    test(a2, 5, 5, 30);
    test(a2, 5, 6, 23);

    return 0;
}

File: Array/875.py
class Solution:
    def minEatingSpeed(self, piles, h):
        # TODO: 你寫
        left = 1
        '''
        max = 0
        for i in range (n):
            if piles[i] > max:
                max = piles[i]
        '''
        max_piles = max(piles)
        right = max_piles
        #區間縮到只剩 1 個
        #找邊界 不是找值
        while left < right:
            mid = (left+right)//2
            time = 0
            for p in piles:
                time += (p+mid-1)//mid
            #太慢
            if time > h:
                left = mid+1
            #太快
            else:
                #最小可行
                right = mid
                #最大可行
                #mid = (l+r+1)//2  left = mid
        return left


def run_test(piles, h, expected):
    sol = Solution()
    ans = sol.minEatingSpeed(piles, h)
    print(f"piles={piles}, h={h}")
    print(f"  ans={ans}, expected={expected}  --> {'PASS' if ans == expected else 'FAIL'}")
    if ans != expected:
        raise AssertionError("Test failed")


if __name__ == "__main__":
    run_test([3,6,7,11], 8, 4)
    run_test([30,11,23,4,20], 5, 30)
    run_test([30,11,23,4,20], 6, 23)
    run_test([1,1,1,1], 4, 1)
    print("All tests passed!")